platform/solr/solr-factory/src/test/groovy/org/codice/solr/factory/impl/SolrClientAdapterAsyncSpec.groovy

```diff
@@ -0,0 +1,1253 @@
+  }
+
+  @Unroll
+  def 'test client #test_client_how where an API call also fails to connect after the initial failure'() {
+    given: "a Solr client that fails direct pings"
+      def client = Mock(SolrClient) {
+        // in case the real client is consulted for the first API call
+        // since either we will be rejected right away if the client is not created but if it is
+        // a ping will be issued to the client
+        ping() >> pingFailResponse
+      }
+
+    and: "controllers that mocks every execution/attempts"
+      def createController = new FailsafeController('SolrClient Creation').onNextExecution {
+        waitTo('create').onlyIf(blocking_create)
+            .then().doReturn(client)
+      }
+      def pingController = new FailsafeController('SolrClient Ping').onNextExecution {
+        waitTo('ping').onlyIf(blocking_ping)
+            .then().doReturn()
+      }
+
+    when: "creating an adapter"
+      def adapter = new SolrClientAdapter(CORE, creator, createController.&with, pingController.&with)
+
+    then: "the adapter should be unavailable because it is being initialized"
+      !adapter.available
+
+    when: "calling any api methods"
+      adapter.request(request, COLLECTION)
+
+    then: "the method should not have been delegated to the client"
+      0 * client.request(request, COLLECTION)
+
+    and: "verify it throws an unavailable exception indicating we are initializing"
+      def e = thrown(UnavailableSolrException)
+
+      e.message.contains("initializing '$CORE' client")
+
+    when: "triggering the client creation/connection"
+      createController.notifyTo('create')
+      pingController.notifyTo('ping')
+
+    and: "waiting for failsafe to complete"
+      createController.waitForSuccessfulCompletion()
+      pingController.waitForSuccessfulCompletion()
+
+    then: "the adapter should be available"
+      adapter.isAvailable()
+
+    when: "calling any api methods"
+      def returnedResponse = adapter.request(request, COLLECTION)
+
+    then: "verify the underlying client was called"
+      1 * client.request(request, COLLECTION) >> response
+      returnedResponse.is(response)
+
+    and: "the underlying client should never be closed"
+      0 * client.close()
+
+    and: "verify failsafe controllers"
+      createController.verify()
+      pingController.verify()
+
+    cleanup:
+      createController?.shutdown()
+      pingController?.shutdown()
+
+    where:
+      test_client_how                                          || blocking_create | blocking_ping
+      'creation before it becomes available'                   || true            | false
+      'connection as it becomes available'                     || false           | true
+      'before creating and connecting as it becomes available' || true            | true
+  }
+
+  def 'test client connection as it becomes available where an API call succeeds connecting first'() {
```

Comment:
❓Not sure I understand what we're testing here? What does the `waitToBeCancelled()` do on L300 and how does the `pingController` gets past that? What does the `CancellationException` do when it's thrown? Any chance we can improve the description?

Reply:
I added a bit to the method name.
Basically, the design of the adapter is to always try to recover/initialize in the background. However, if someone attempts to use an API method at that time, the API method will attempt to ping on its own even though the background logic is trying to do so too. If the ping from the API call succeeds first, the state of the adapter will be changed and the background logic will be cancelled. 

This is what we are testing here. The ping done in the background is never returning (because of the `waitToBeCancelled`) and a call to any API method will succeed the ping, change the state of the adapter and cancel the failsafe request which would then wakeup that `waitToBeCancelled`. The typical way for Failsafe to respond when this is done is by having the `CancellationException` being thrown back. Since we are mocking Failsafe, we need to reproduce that same condition.